### PR TITLE
Create schema data objects

### PR DESCRIPTION
**Pull request by rymorale**
## Motivation and Context
This is the initial Internal refactor pr to replace the `Inbound` class with specific schema data objects to support schema based consequence details.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/FeedItemSchemaData.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    private int expiryDate;
+    private Map<String, Object> meta;
+
+    FeedItemSchemaData(final JSONObject schemaData) {
+        try {
+            final JSONObject consequenceDetails = InternalMessagingUtils.getConsequenceDetails(schemaData);
+            final Map<String, Object> feedItemData = JSONUtils.toMap(consequenceDetails.getJSONObject(DATA));
+            this.contentType = ContentType.fromString(DataReader.getString(feedItemData, CONTENT_TYPE));
+            if (contentType.equals(ContentType.APPLICATION_JSON)) {
+                this.content = DataReader.getTypedMap(Object.class, feedItemData, CONTENT);
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/FeedItemSchemaData.java:54]: Is this always guaranteed to be a `JSONObject`? Any chance we might get `JSONArray`?

Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/FeedItemSchemaData.java:54]: Will the feed UI allows for custom content or will it be only OOTB Feed Item?

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/InAppSchemaData.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    private Map<String, Object> meta;
+    private Map<String, Object> mobileParameters;
+    private Map<String, Object> webParameters;
+    private List<String> remoteAssets = new ArrayList<>();
+
+    InAppSchemaData(final JSONObject schemaData) {
+        try {
+            this.contentType = ContentType.fromString(schemaData.optString(CONTENT_TYPE));
+            if (contentType.equals(ContentType.APPLICATION_JSON)) {
+                this.content = JSONUtils.toMap(schemaData.getJSONObject(CONTENT));
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/InAppSchemaData.java:51]: Same question here. Can this be a `JSONArray`? More importantly this cannot be anything other than HTML considering it is constructed by AJO. Should we allow support for other content types?

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java [hunk 3]
```diff
@@ -113,82 +100,137 @@ void track(final PropositionEventType interaction) {
...
     /**
-     * Creates an {@code Inbound} object from this {@code PropositionItem}'s content.
+     * Creates a {@code MessagingPropositionItem} object from the provided {@code RuleConsequence}.
      *
-     * @return {@link Inbound} object created from this {@link MessagingPropositionItem}'s content.
+     * @return {@link MessagingPropositionItem} object created from the provided {@link RuleConsequence}.
      */
-    public Inbound decodeContent() {
-        Inbound inboundContent = null;
+    public static MessagingPropositionItem fromRuleConsequence(final RuleConsequence consequence) {
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java:107]: This method returns the same type of object as the class it is in. Can we refactor it to be constructor instead?
`public MessagingPropositionItem(final RuleConsequence consequence) {
              // get uniqueId, schema and itemData from consequence
              this(uniqueId, schema, itemData);
}`

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java [hunk 4]
```diff
@@ -113,82 +100,137 @@ void track(final PropositionEventType interaction) {
...
     }
 
     /**
-     * Creates a {@code PropositionItem} object from the provided {@code Map<String, Object>}.
+     * Creates a schema data object from this {@code MessagingPropositionItem}'s content.
+     *
+     * @param schemaType {@link SchemaType} to be used when creating the {@link T} object.
+     * @return {@code T} object created from the provided {@link MessagingPropositionItem}'s content.
+     */
+    private <T> T createSchemaData(final SchemaType schemaType) {
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java:182]: I would recommend having all schema data types extend from a common interface/abstract class. Would make the code more extensible to new types in the future

Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java:182]: We can use Curiously Recurring Template Pattern here, similar to what Core has done with Presentation types in the new Jetpack Compose refactor
```
public abstract class SchemaData<T extends SchemaData<T>>{
                abstract Object getContent();
}

public class HtmlContentSchemaData extends SchemaData<HtmlContentSchemaData> {
             public Object getContent() {
                      //HTML specific implementation for getContent
             }
}


public class FeedItemSchemaData extends SchemaData<FeedItemSchemaData> {
             public Object getContent() {
                      //Feed specific implementation for getContent
             }
}

public class InAppSchemaData extends SchemaData<InAppSchemaData> {
             public Object getContent() {
                      //Feed specific implementation for getContent
             }
}

private <T> T createSchemaData(final SchemaType schemaType) {
           // all switch cases, no need to cast
            case HTML_CONTENT:
                      return (T) new HtmlContentSchemaData(ruleJson);
}
```

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java [hunk 5]
```diff
@@ -113,82 +100,137 @@ void track(final PropositionEventType interaction) {
...
-                    } else if (contentObject instanceof List) {
-                        jsonArray = new JSONArray((List) contentObject);
-                    } else {
-                        jsonContent = new JSONObject(contentObject.toString());
-                    }
-                }
-            } else { // html or text content
-                content = DataReader.getString(contentMap, PAYLOAD_CONTENT);
-            }
+            final String uniqueId = DataReader.getString(eventData, ID);
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java:216]: Same suggestion here, how about making this a constructor
```
public MessagingPropositionItem(final Map<String, Object> eventData) {
           // get uniqueId, schema and itemData from eventData
           this(uniqueId, schema, itemData); 
}
```

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/FeedItemSchemaData.java (hunk 1 above)
Comment (rymorale) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/FeedItemSchemaData.java:54]: in-app and feed items are always delivered via a ruleset schema which each rule being a JSON object. For now only OOTB feed items are supported.

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/InAppSchemaData.java (hunk 2 above)
Comment (rymorale) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/InAppSchemaData.java:51]: like feed items, in-app are JSON objects.

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java (hunk 3 above)
Comment (rymorale) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java:107]: i think I prefer to keep the static method as there may be some confusion if i use two different constructors for the object.

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java (hunk 4 above)
Comment (rymorale) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java:182]: good point, i'll make this change

**Pull request merged**